File: Java/com/leetcode/linkedListCycleII/LinkedListCycleII.java
package com.leetcode.linkedListCycleII;

import com.leetcode.dataStructures.ListNode;

public class LinkedListCycleII {
    public ListNode detectCycle(ListNode head) {
        if(head == null || head.next == null)
            return null;
        if(head.next == head)
            return head;
        
        ListNode slow = head;
        ListNode fast = head;
        boolean hasCircle = false;
        while(fast.next != null){
            slow = slow.next;
            fast = fast.next;
            if(fast.next == null){
                return null;
            }
            fast = fast.next;
            if(slow == fast){
                hasCircle = true;
                break;
            }
        }
        if(!hasCircle)
            return null;
        slow = head;
        if(head == fast)
            return head;
        while(true){
            slow = slow.next;
            fast = fast.next;
            if(slow == fast){
                return slow;
            }
        }
    }
}

File: Python/linked_list_cycle_ii/linked_list_cycle_ii.py
class Solution(object):
    def detectCycle(self, head):
        p = head
        q = head
        while p is not None and q is not None:
            p = p.next
            q = q.next
            if p is None or q is None:
                return None
            q = q.next
            if q is None:
                return None
            if p == q:
                break

        if p is None or q is None:
            return None

        p = head
        while p != q:
            p = p.next
            q = q.next

        return p
